### DISCUSSION TITLE
Using createPersister for mutations

### DISCUSSION THREAD

**Discussion by davidjbng** [TanStack/query#6439] [Ideas]
Allow persisting mutation state through `createMutationPersister`.

## Idea

```ts
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      persister: experimental_createQueryPersister({ 
	    // ... 
	  }),
    },
    mutations: {
	  // 👇 new persister option
      persister: experimental_createMutationPersister({ 
	    // ... 
	  }),
    },
  },
})
```

### Usage

- Mutations need a `mutationKey` to be persisted
- Persisted mutations can be accessed through `useMutationState`

```ts
useMutation({
   mutationKey: ['key']
   // ...
})

const persistedMutations = useMutationState({
   filters: {
      mutationKey: ['key']
   }
})
```

### Resuming Mutations

Pending Mutations could be resumed automatically when a mutation with a matching key mounts.

## Use cases

- When you are doing optimistic ui with `mutation.variables`. You want to persist failed mutation so the user can retry them at any point even after reloading the app.

**Comment by TkDodo** [comment]
It's a good idea but we have to think this through. One problem is that with mutations, they are never "picked up" when a `useMutation` mounts, because every call to `useMutation` yields a new mutation. Even if we get that to work, it also would be weird if mutations don't resume when the app mounts, because with this approach, they would only resume when the component that triggered the mutation is rendered...

@DamianOsipiuk fyi

**Reply by davidjbng** [reply]
Maybe mutations do not necessarily need to be resumed automatically, they should just persist the state I can access through `useMutationState` ~~or `useMutation`~~ via `mutationKey`.
Edit: I see that hydrating `useMutation` might be different, but `useMutationState` seems possible

This would solve my specific use case:
I'm doing some optimistic UI using `mutation.variables`. When a mutation fails, I display a retry button alongside. I specifically want to keep `mutation.variables` around as long as the mutation has not succeeded, so the user can reload the app and still see all failed mutations to retry.

It might be possible build this feature some other way - basically by persisting pending mutations myself - but it feels like it would be solved if the mutation state was persisted in some way.

**Reply by TkDodo** [reply]
oh I see - yeah it would very likely work with `useMutationState` if we'd just persist it. I think we could also get some sort of "automatic" resume, e.g. to resume all paused mutations that match a key when the same key mounts? Not sure if there's a use-case for that though...

**Reply by davidjbng** [reply]
So do you think it should be done via a `persister` option for mutations like this?

```ts
const persister = experimental_createPersister({ 
	// ... 
}),
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      persister
    },
    mutations: {
	  // 👇 new persister option for mutations
      persister
    }
  },
})
```

I'm willing to work this and try it out. Is there something good to know or do you have any more pointers?
First I would only focus on persisting mutation state and making it available through `useMutationState`.

**Reply by TkDodo** [reply]
yes this would be the proposed api. Since the whole thing is experimental - feel free to hack away :)

**Reply by DamianOsipiuk** [reply]
I believe we could make this work with a couple of prerequisites.
- every mutation that would need to be persisted, would need to have `mutationKey` to properly match it to the usage place
- persist only mutations that were not successful, or maybe only `paused` ones?
- we would need an iterator over the storage entries in case we would like to restore all of them at once. And some kind of unique prefix for mutations. There is already [resumePausedMutations](https://github.com/TanStack/query/blob/main/packages/query-core/src/mutationCache.ts#L163) function which we could hook up.

**Reply by davidjbng** [reply]
After looking through the implementation of `createPersister` it seems like mutation persister should be its own function `createMutationPersister`. This leads to the question should `createPersister` be renamed to `createQueryPersister`?

The api would look more like this
```ts
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      persister: experimental_createQueryPersister({ 
	    // ... 
	  }),
    },
    mutations: {
      persister: experimental_createMutationPersister({ 
	    // ... 
	  }),
    },
  },
})
```

## Why separate `createMutationPersister`?

- The persister function for mutations will most likely have a different interface. A separate function makes this way easier to type
- The implementation details will be different to `createPersister` (e.g. mutations cannot be stale, prob some other logic)

**Reply by DamianOsipiuk** [reply]
Yeah, i also think a different function would be more appropriate for this

**Reply by TkDodo** [reply]
> persist only mutations that were not successful, or maybe only paused ones?

yes, we only persist paused mutations per default in the PersistProvider, too:

https://github.com/TanStack/query/blob/497804ee00927da14ee8ff7f8218f270def7a91e/packages/query-core/src/hydration.ts#L68-L70

> we would need an iterator over the storage entries in case we would like to restore all of them at once.

this would also be great for the retries we discussed, in case of storage being full ...
